File: src/2023-04-13-custom_rpc_options/python/main.py
from proto.world_pb2 import DESCRIPTOR

def get_option_value(opts, id):
  for field in opts.ListFields():
    (desc, value) = field

    if value != "" and desc.name == id:
      return value

def get_service_method(sd, predicate):
  for method_name in sd.methods_by_name:
    md = sd.methods_by_name[method_name]

    if predicate(md):
      return md

def get_method_option_value(sd, id):
  md = get_service_method(sd, lambda md: get_option_value(md.GetOptions(), id) != None)

  return get_option_value(md.GetOptions(), id)

def get_method_extension(fd, id):
  for svc_name in fd.services_by_name:
    sd = fd.services_by_name[svc_name]
    value = get_method_option_value(sd, id)

    if value != None:
      return value

print(get_method_extension(DESCRIPTOR, "hello"))

File: src/2023-04-13-custom_rpc_options/csharp/main.cs
#nullable enable annotations

using System;
using System.Linq;
using System.Collections.Generic;
using pb = global::Google.Protobuf;

static class HelloWorldApp {

	static private T GetOptionValue<T>(
		this pb.Reflection.MethodDescriptor md,
		pb::Extension<pb.Reflection.MethodOptions, T> id
	) => md.GetOptions().GetExtension(id);

	static private IEnumerable<pb.Reflection.MethodDescriptor> GetServiceMethod(
		this IEnumerable<pb.Reflection.ServiceDescriptor> services,
		Func<pb.Reflection.MethodDescriptor, bool> predicate
	) => from svc in services
			 from method in svc.Methods
			 where predicate(method)
			 select method;

	static private T GetMethodOptionValue<T>(
		this pb.Reflection.FileDescriptor fd,
		pb::Extension<pb.Reflection.MethodOptions, T> id
	) => fd.Services
				 .GetServiceMethod(md => md.GetOptionValue(id) != null)
				 .FirstOrDefault()
				 .GetOptionValue(id);

	static public void Main(String[] args)
	{
		var id = HelloExtensions.Hello;
		string world = WorldReflection.Descriptor.GetMethodOptionValue(id);

		if (world.Length != 0)
			System.Console.WriteLine(world);
	}
}